specs/_features/epbs/beacon-chain.md

```diff
@@ -0,0 +1,779 @@
+    signing_root = compute_signing_root(signed_envelope.message, get_domain(state, DOMAIN_BEACON_BUILDER))
+    return bls.Verify(builder.pubkey, signing_root, signed_envelope.signature)
+```
+
+#### New `process_execution_payload_header`
+
+```python
+def process_execution_payload_header(state: BeaconState, signed_header: SignedExecutionPayloadHeader) -> None:
+    assert verify_execution_payload_header_signature(state, signed_header)
+    header = signed_header.message
+    # Verify consistency of the parent hash with respect to the previous execution payload header
+    assert header.parent_hash == state.latest_execution_payload_header.block_hash
+    # Verify prev_randao
+    assert header.prev_randao == get_randao_mix(state, get_current_epoch(state))
+    # Verify timestamp
+    assert header.timestamp == compute_timestamp_at_slot(state, state.slot)
+    # Cache execution payload header
+    state.current_signed_execution_payload_header = signed_header
+```
+
+#### Modified `process_execution_payload`
+*Note*: `process_execution_payload` is now an independent check in state transition. It is called when importing a signed execution payload proposed by the builder of the current slot.
+
+```python
+def process_execution_payload(state: BeaconState, signed_envelope: SignedExecutionPayloadEnvelope, execution_engine: ExecutionEngine) -> None:
+    # Verify signature [New in ePBS]
+    assert verify_execution_envelope_signature(state, signed_envelope)
+    payload = signed_envelope.message.payload
+    # Verify consistency with the committed header
+    hash = hash_tree_root(payload)
+    previous_hash = hash_tree_root(state.current_signed_execution_payload_header.message)
+    assert hash == previous_hash
+    # Verify the execution payload is valid
+    inclusion_list = execution_engine.verify_and_notify_new_payload(NewPayloadRequest(execution_payload=payload))
+    # Verify and update the proposers inclusion lists
+    update_tx_inclusion_lists(state, payload, inclusion_list)
+    # Process Withdrawals in the payload
+    process_withdrawals(state, payload)
+    # Cache the execution payload header
+    state.latest_execution_payload_header = state.current_signed_execution_payload_header.message 
+    # Verify the state root
+    assert signed_envelope.message.state_root == hash_tree_root(state)
+```
+
+#### New `process_tx_inclusion_list`
+
+```python
+def process_tx_inclusion_list(state: BeaconState, block: BeaconBlock, execution_engine: ExecutionEngine) -> None:
+    inclusion_list = block.body.tx_inclusion_list
+    # Verify that the list is empty if the parent consensus block did not contain a payload
+    if state.current_signed_execution_payload_header.message != state.latest_execution_payload_header:
+        assert not inclusion_list
+        return
+    assert notify_new_inclusion_list(execution_engine, inclusion_list)
+    state.previous_tx_inclusion_list = state.current_tx_inclusion_list
```

Comment:
Something smells here. If there's left over IL from the previous slot, we'll lose them. I think it should be an append, but there are edge cases like out of bound / gas to consider. It's an annoying problem